code/include/playfab/internal_memory.h

```diff
@@ -0,0 +1,163 @@
+#pragma once
+
+#include <memory>
+#include <unordered_map>
+#include <set>
+#include <queue>
+#include <unordered_set>
+
+namespace PlayFab
+{
+
+//------------------------------------------------------------------------------
+// Allocator
+//------------------------------------------------------------------------------
+template<class T>
+struct Allocator
+{
+    using value_type = T;
+    using is_always_equal = std::true_type;
+
+    Allocator() = default;
+    template<class U>
+    Allocator(Allocator<U> const&) {}
+
+    T* allocate(size_t n);
+    void deallocate(_In_opt_ void* p, size_t) noexcept;
+};
+
+template<class T>
+bool operator==(Allocator<T> const&, Allocator<T> const&);
+template<class T>
+bool operator!=(Allocator<T> const&, Allocator<T> const&);
+
+//------------------------------------------------------------------------------
+// Deleter
+//------------------------------------------------------------------------------
+template<class T>
+struct Deleter
+{
+    void operator()(T* ptr) const noexcept;
+};
+
+//------------------------------------------------------------------------------
+// Smart Pointers
+//------------------------------------------------------------------------------
+template<class T>
+using SharedPtr = std::shared_ptr<T>;
+
+template<class T>
+using WeakPtr = std::weak_ptr<T>;
+
+template<class T, class TDeleter = Deleter<T>>
+using UniquePtr = std::unique_ptr<T, TDeleter>;
+
+//------------------------------------------------------------------------------
+// Factories
+//------------------------------------------------------------------------------
+template<class T, class... TArgs>
+SharedPtr<T> MakeShared(TArgs&&... args);
+
+template<class T, class... TArgs>
+UniquePtr<T> MakeUnique(TArgs&&... args);
+
+//------------------------------------------------------------------------------
+// Template implementations
+//------------------------------------------------------------------------------
+template<class T>
+T* Allocator<T>::allocate(size_t n)
+{
+    if (n > (static_cast<size_t>(-1) / sizeof(T)))
+    {
+        throw std::bad_alloc{};
+    }
+
+    auto rawPtr = malloc(n * sizeof(T)); // Bug 15586224: how to ensure alignment?
+    return static_cast<T*>(rawPtr);
+}
+
+template<class T>
+void Allocator<T>::deallocate(_In_opt_ void* p, size_t) noexcept
+{
+    free(p);
+}
+
+template<class T>
+bool operator==(Allocator<T> const&, Allocator<T> const&)
+{
+    return true;
+}
+
+template<class T>
+bool operator!=(Allocator<T> const&, Allocator<T> const&)
+{
+    return false;
+}
+
+template<class T>
+void Deleter<T>::operator()(T* ptr) const noexcept
+{
+    ptr->~T(); // destroy
+    Allocator<T> alloc{};
+    alloc.deallocate(ptr, 1);
+}
+
+template<class T, class... TArgs>
+SharedPtr<T> MakeShared(TArgs&&... args)

```

Comment:
Also... defined twice?

Reply:
I see a declaration above but only one definition. I guess its split for clarity, but I just copied this file :)

---
In reply to: [590890806](https://github.com/jplafonta/XPlatCppSdk/pull/2#discussion_r590890806) [](ancestors = 590890806)